src/Symfony/Component/HttpFoundation/HeaderBag.php

```diff
@@ -68,10 +75,21 @@ public function __toString(): string
     public function all(?string $key = null): array
     {
         if (null !== $key) {
-            return $this->headers[strtr($key, self::UPPER, self::LOWER)] ?? [];
+            $uniqueKey = strtr($key, self::UPPER, self::LOWER);
+            if (str_ends_with($uniqueKey, 'cache-control')) {
+                return [$this->getCacheControl($this->extractCacheControlTarget($key))->getCacheControlHeader()];
+            }
+
+            return $this->headers[$uniqueKey] ?? [];
         }
 
-        return $this->headers;
+        $headers = $this->headers;
+        // edge case: what if extending class directly changed Cache-Control in the $headers array?
```

Comment:
should we check for the existence of the header we are about to create and not set it in that case?

Reply:
also, should we check if the instruction set is empty to not create an empty header, or should we do as we do now and set the emtpy header with no value? the presence of the header could be used by something downstream even if there are no values.